### PR TITLE
Lazy import cv2 only when running reference for ImageDecoder op (#5491)

### LINKED ISSUE #5491: collect_snippets requires cv2 dependency not available on macOS
# Bug Report

### Describe the bug
#5294 added a `cv2` dependency to `onnx.backend.test.case.collect_snippets()` which is not available on macOS. 

### System information
- OS Platform and Distribution (*e.g. Linux Ubuntu 20.04*): macOS 13.4
- ONNX version (*e.g. 1.13*): main
- Python version: 3.11.4

### Reproduction instructions
```
~/: python -m venv onnx-env  
~/: source ./onnx-env/bin/activate
(onnx-env) ~/: git clone --quiet --recursive https://github.com/onnx/onnx.git
(onnx-env) ~/: pip install ./onnx
(onnx-env) ~/: python
Python 3.11.4 (main, Jun 20 2023, 17:23:00) [Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import onnx.backend.test.case
>>> onnx.backend.test.case.collect_snippets()
  File "~/onnx-env/lib/python3.11/site-packages/onnx/reference/ops/op_image_decoder.py", line 6, in <module>
    import cv2
ModuleNotFoundError: No module named 'cv2'
>>> ^D
(onnx-env) ~/: pip install cv2
ERROR: Could not find a version that satisfies the requirement cv2 (from versions: none)
ERROR: No matching distribution found for cv2
```

### Expected behavior
`collect_snippets` should succeed without error if `cv2` is not available, for example by delaying load?

### Notes

@jantonguirao @xadupre @gramalingam

### PR DESCRIPTION
**Pull request by lutzroeder**
### Description
Fix lazy import of `cv2` per discussion in #5491.

### Motivation and Context
Fixes #5491. Follow-up to #5492.

@jcwchen @jantonguirao @xadupre @gramalingam

### CODE REVIEW

**Review by jcwchen — APPROVED** [review]
Previously I thought cv2 was only used when running reference  and I didn't notice that cv2 was imported during test generation also... Thank you for the PR!

To confirm, will this PR help resolve your issue? If your usage involves test generation, it might hit cv2 not found issue again.

File: onnx/backend/test/case/node/image_decoder.py [hunk 1]
```diff
@@ -40,6 +39,13 @@ def generate_checkerboard(width, height, square_size):
 
 
 def generate_test_data(extension, pixel_format="RGB", h=40, w=40, tile_sz=5):
+    try:
+        # pylint: disable=import-outside-toplevel`
```
Comment (justinchuby) [inline onnx/backend/test/case/node/image_decoder.py:43]:
```suggestion
        # pylint: disable=import-outside-toplevel
```

nit

**Comment by lutzroeder** [comment]
> Previously I thought cv2 was only used when running reference and I didn't notice that cv2 was imported during test generation also... Thank you for the PR and catching this!
> To confirm, will this PR help resolve your issue? If your usage involves test generation, it might hit cv2 not found issue again.

Good question. It looks like `collect_snippets()` invokes this code path so it would not fix the issue. Any ideas why this code is being executed during import?

```
    snippets = onnx.backend.test.case.collect_snippets()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/env/onnx/lib/python3.11/site-packages/onnx/backend/test/case/__init__.py", line 13, in collect_snippets
    import_recursive(sys.modules[__name__])
  File "~/env/onnx/lib/python3.11/site-packages/onnx/backend/test/case/utils.py", line 45, in import_recursive
    import_recursive(module)
  File "~/env/onnx/lib/python3.11/site-packages/onnx/backend/test/case/utils.py", line 43, in import_recursive
    module = importlib.import_module(module_name)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "~/env/onnx/lib/python3.11/site-packages/onnx/backend/test/case/node/image_decoder.py", line 66, in <module>
    class ImageDecoder(Base):
  File "~/env/onnx/lib/python3.11/site-packages/onnx/backend/test/case/base.py", line 41, in __init__
    export()
  File "~/env/onnx/lib/python3.11/site-packages/onnx/backend/test/case/node/image_decoder.py", line 76, in export_image_decoder_decode_jpeg_rgb
    data, output = generate_test_data(".jpg", "RGB")
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/env/onnx/lib/python3.11/site-packages/onnx/backend/test/case/node/image_decoder.py", line 46, in generate_test_data
    raise ImportError(
ImportError: opencv-python must be installed to use the reference implementation of the ImageDecoder operator
```

**Review by xadupre — APPROVED** [review]

**Comment by jcwchen** [comment]
Ideally test generation should be as simple as possible without other new dependency especially that dependency was not used in other places (reference will only use cv2 when needed). Let's merge this PR first and later I will have another PR to freeze input/output as a pure NumPy array and simply keep generate_checkerboard/generate_test_data for future reference.

@lutzroeder could you please fix [DCO issue](https://github.com/onnx/onnx/blob/main/CONTRIBUTING.md#dco) for merging this PR if you have time? Or, I can also cover this PR in mine. Thank you!

**Review by gramalingam — APPROVED** [review]

**Pull request closed without merging**